### PR TITLE
Keeps Object Storage and Postgres in Sync

### PR DESCRIPTION
**Pull request by brendoncarroll**
These changes let us create and delete chunks without ever ending up with a chunk without a corresponding object in object storage.  It is possible to wind up with multiple objects per chunk, and objects without a Postgres entry, but the existence of these anomalies will not impact correctness, and there are background services to eliminate them eventually.

### CODE REVIEW

File: src/internal/storage/chunk/client.go [hunk 1]
```diff
@@ -43,73 +50,105 @@ func NewClient(store kv.Store, mdstore MetadataStore, tr track.Tracker, name str
...
 	chunkID := Hash(chunkData)
 	var pointsTo []string
 	for _, cid := range md.PointsTo {
-		pointsTo = append(pointsTo, ObjectID(cid))
+		pointsTo = append(pointsTo, cid.TrackerID())
 	}
-	chunkOID := ObjectID(chunkID)
-	if err := dbutil.WithTx(ctx, c.mdstore.DB(), func(tx *sqlx.Tx) error {
-		if err := c.mdstore.SetTx(tx, chunkID, md); err != nil && err != ErrMetadataExists {
+	chunkOID := chunkID.TrackerID()
```
Comment (brycemcanally) [inline src/internal/storage/chunk/client.go:62]: We should probably use `TID` as the suffix for tracker ID to avoid confusion between object storage objects and tracker objects. This is pretty minor, but I think it might be important for readability.

File: src/internal/storage/chunk/gc.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+func (gc *GarbageCollector) deleteOne(ctx context.Context, ent Entry) error {
+	if err := gc.deleteObject(ctx, ent.ChunkID, ent.Gen); err != nil {
+		return err
+	}
+	return gc.deleteEntry(ctx, ent.ChunkID, ent.Gen)
+}
+
+func (gc *GarbageCollector) deleteObject(ctx context.Context, chunkID ID, gen uint64) error {
+	return gc.s.store.Delete(ctx, chunkKey(chunkID, gen))
```
Comment (brycemcanally) [inline src/internal/storage/chunk/gc.go:84]: I think we could run into issues if we return an error here when the object does not exist. For example, if the machine crashes immediately after this call and before deleting the entry, then I think the next garbage collector would be stuck in a loop once it reached this point again for the object.

File: src/internal/storage/chunk/gc.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func (gc *GarbageCollector) deleteEntry(ctx context.Context, chunkID ID, gen uint64) error {
+	_, err := gc.s.db.ExecContext(ctx, `
+	DELETE FROM storage.chunk_objects
+	WHERE chunk_id = $1 AND gen = $2 AND tombstone = TRUE
+	LIMIT 1
```
Comment (brycemcanally) [inline src/internal/storage/chunk/gc.go:91]: Why is this necessary?

File: src/internal/storage/chunk/gc.go (hunk 2 above)
Comment (brendoncarroll) [reply src/internal/storage/chunk/gc.go:84]: I think delete is idempotent.  If it's not at the object level; I propose changing it at the kv level.

File: src/internal/storage/chunk/gc.go (hunk 3 above)
Comment (brendoncarroll) [reply src/internal/storage/chunk/gc.go:91]: It's not.  Would you like to join the add limits to point deletes cargo cult?

File: src/internal/storage/chunk/client.go (hunk 1 above)
Comment (brendoncarroll) [reply src/internal/storage/chunk/client.go:62]: Yeah this makes sense.

File: src/internal/storage/chunk/gc.go (hunk 2 above)
Comment (brendoncarroll) [reply src/internal/storage/chunk/gc.go:84]: Okay so delete is not idempotent on GCS
https://cloud.google.com/storage/docs/xml-api/delete-object

But it is in AWS.
https://docs.aws.amazon.com/sdkfornet1/latest/apidocs/html/M_Amazon_S3_AmazonS3Client_DeleteObject.htm

So I changed my mind, this needs to be uniform before it leaves the obj package.

File: src/internal/storage/chunk/client.go [hunk 4]
```diff
@@ -43,73 +50,105 @@ func NewClient(store kv.Store, mdstore MetadataStore, tr track.Tracker, name str
...
+func (c *trackedClient) Get(ctx context.Context, chunkID ID, cb kv.ValueCallback) (retErr error) {
+	var gen uint64
+	err := c.db.Get(&gen, `
+	SELECT gen
+	FROM storage.chunk_objects
+	WHERE uploaded = TRUE AND tombstone = FALSE AND chunk_id = $1
+	LIMIT 1
+	`, chunkID)
+	if err != nil {
+		err = errors.Errorf("no objects for chunk %v", chunkID)
```
Comment (brycemcanally) [inline src/internal/storage/chunk/client.go:123]: It seems like this should be handling sql no rows and just returning the error otherwise.

File: src/internal/storage/chunk/gc.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	`, chunkID, gen)
+	return err
+}
+
+// ReduceObjectsPerChunk eliminates redundant objects
+func ReduceObjectsPerChunk(ctx context.Context, db *sqlx.DB) error {
```
Comment (brycemcanally) [inline src/internal/storage/chunk/gc.go:97]: I think we should punt on these two functions for now. They seem pretty low leverage since we expect this issue to be pretty rare. It looks like you are doing that anyways since they are not in use outside of a test.

File: src/internal/storage/chunk/gc.go (hunk 2 above)
Comment (brycemcanally) [reply src/internal/storage/chunk/gc.go:84]: Okay, so are you thinking we should have a standard error for this case or just requiring that the clients don't return an error if the object does not exist?

File: src/internal/storage/chunk/gc.go (hunk 3 above)
Comment (brycemcanally) [reply src/internal/storage/chunk/gc.go:91]: lol, no thanks.

File: src/internal/storage/chunk/gc.go (hunk 2 above)
Comment (brendoncarroll) [reply src/internal/storage/chunk/gc.go:84]: I like no error. I think that is more useful generally. I can't think of a use case where we care if there was something there.

Comment (brendoncarroll) [reply src/internal/storage/chunk/gc.go:84]: Okay, for the sake of this P.R. lets assume this is the behavior. And track it here:
https://github.com/pachyderm/pachyderm/pull/5866#issuecomment-804299195

**Review by brycemcanally — APPROVED** [review]
lgtm
